overlord/snapshotstate/backend/backend.go

```diff
@@ -445,6 +460,177 @@ func addDirToZip(ctx context.Context, snapshot *client.Snapshot, w *zip.Writer,
+	return osutil.FileExists(t.importInProgressFilepath())
+}
+
+// Cancel cancels a snapshot import and cleanups any files on disk belonging
+// to this snapshot ID.
+func (t *importTransaction) Cancel() error {
+	if t.committed {
+		return ErrCannotCancel
+	}
+	inProgressImports, err := filepath.Glob(t.importInProgressFilesGlob())
+	if err != nil {
+		return err
+	}
+	var errs []error
+	for _, p := range inProgressImports {
+		if err := os.Remove(p); err != nil {
+			errs = append(errs, err)
+		}
+	}
+	if len(errs) > 0 {
+		buf := bytes.NewBuffer(nil)
+		for _, err := range errs {
+			fmt.Fprintf(buf, " - %v\n", err)
+		}
+		return fmt.Errorf("cannot cancel import of id %d:\n%s", t.id, buf.String())
+	}
+	return nil
+}
+
+// Commit will commit a given transaction
+func (t *importTransaction) Commit() error {
+	if err := os.Remove(t.importInProgressFilepath()); err != nil {
+		return err
+	}
+	t.committed = true
+	return nil
+}
+
+// Import a snapshot from the export file format
+func Import(ctx context.Context, id uint64, r io.Reader) (snapNames []string, err error) {
+	errPrefix := fmt.Sprintf("cannot import snapshot %d", id)
+
+	tr := newImportTransaction(id)
+	if tr.InProgress() {
+		return nil, fmt.Errorf("%s: already in progress for this id", errPrefix)
+	}
+	if err := tr.Start(); err != nil {
+		return nil, err
+	}
+	defer tr.Cancel()
+
+	// Unpack and validate the streamed data
+	snapNames, err = unpackVerifySnapshotImport(r, id)
+	if err != nil {
+		return nil, fmt.Errorf("%s: %v", errPrefix, err)
+	}
+	if err := tr.Commit(); err != nil {
+		return nil, err
+	}
+
+	return snapNames, nil
+}
+
+func unpackVerifySnapshotImport(r io.Reader, realSetID uint64) (snapNames []string, err error) {
+	var exportFound bool
+
+	tr := tar.NewReader(r)
+	var tarErr error
+	var header *tar.Header
+
+	for tarErr == nil {
+		header, tarErr = tr.Next()
+		if tarErr == io.EOF {
+			break
+		}
+		switch {
+		case tarErr != nil:
+			return nil, fmt.Errorf("failed reading snapshot import: %v", tarErr)
+		case header == nil:
+			// should not happen
+			return nil, fmt.Errorf("tar header not found")
+		case header.Typeflag == tar.TypeDir:
+			return nil, errors.New("unexpected directory in import file")
+		}
+
+		if header.Name == "export.json" {
+			// XXX: read into memory and validate once we
+			// hashes in export.json
+			exportFound = true
+			continue
+		}
+
+		// Format of the snapshot import is:
+		//     $setID_.....
+		// But because the setID is local this will not be correct
+		// for our system and we need to discard this setID.
+		//
+		// So chop off the incorrect (old) setID and just use
+		// the rest that is still valid.
+		l := strings.SplitN(header.Name, "_", 2)
+		if len(l) != 2 {
+			return nil, fmt.Errorf("unexpected filename in stream: %v", header.Name)
+		}
+		targetPath := path.Join(dirs.SnapshotsDir, fmt.Sprintf("%d_%s", realSetID, l[1]))
+
+		t, err := os.OpenFile(targetPath, os.O_CREATE|os.O_RDWR, 0600)
```

Comment:
this copy should probably be its own helper because the defer t.Close() is scoped to the function and that is slightly awkward to me. Be we do use defer inside for loops but is not great in general I think